### ISSUE TITLE
Separate update event into the update, soft remove and restore events

### ISSUE THREAD

**Issue by spotykatch** [typeorm/typeorm#8398]
## Feature Description
Separate update event into the update, soft remove and restore events.

### The Problem

Soft remove and recover fire update event, so you have to send some data for subscribers to understand it is not an update.  While listeners can't get the difference at all. So you have to do it like so
```ts
await repository.softRemove(entity, { data: { action: 'soft-remove' } });
```
```ts
afterUpdate(event: UpdateEvent<Entity>): void {
  const { entity, queryRunner: { data: { action } } } = event;
  if ( action === 'soft-remove') {
    // do stuff
  }
}
```
course i.e.
```ts
if (this.softRemoveSubjects.length)
  this.softRemoveSubjects.forEach(subject => this.queryRunner.broadcaster.broadcastAfterUpdateEvent(result, subject.metadata, subject.entity!, subject.databaseEntity, subject.diffColumns, subject.diffRelations));
if (this.recoverSubjects.length)
  this.recoverSubjects.forEach(subject => this.queryRunner.broadcaster.broadcastAfterUpdateEvent(result, subject.metadata, subject.entity!, subject.databaseEntity, subject.diffColumns, subject.diffRelations));
```

### The Solution

Update events should be separated into the update, soft remove, and restore.
```ts
if (this.softRemoveSubjects.length)
  this.softRemoveSubjects.forEach(subject => this.queryRunner.broadcaster.broadcastAfterSoftRemoveEvent(result, subject.metadata, subject.entity!, subject.databaseEntity, subject.diffColumns, subject.diffRelations));
if (this.recoverSubjects.length)
  this.recoverSubjects.forEach(subject => this.queryRunner.broadcaster.broadcastAfterRecoverEvent(result, subject.metadata, subject.entity!, subject.databaseEntity, subject.diffColumns, subject.diffRelations));
```

### Considered Alternatives

OR some data should be sent to listeners and subscribers highlighting it's not an update.

### Additional Context

So the problem is back compatibility. There are some repositories and tons of users that rely on the current functionality. I think it should be discussed.
Maybe there should be an option inside `ormconfig.json`
```json
{
  "separateUpdateEvents": true
}
```
### Are you willing to resolve this issue by submitting a Pull Request?

 - ✖️ Yes, I have the time, and I know how to start.
 - ✅ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by spotykatch** [comment]
LOL, did think that it will take a month or even more. Please, make a review. I'm not sure in some cases.

**Comment by spotykatch** [comment]
Need some help with Chinese docs.

**Comment by shenX-2021** [comment]
@pleerock When the BeforeSoftRemove feature will release?

**Comment by spotykatch** [comment]
It's already released in 0.2.42.

**Comment by shenX-2021** [comment]
@spotykatch thanks!
